// @azure/arm-appplatform/esm/operations/monitoringSettings.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/monitoringSettingsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a MonitoringSettings. */
var MonitoringSettings = /** @class */ (function () {
    /**
     * Create a MonitoringSettings.
     * @param {AppPlatformManagementClientContext} client Reference to the service client.
     */
    function MonitoringSettings(client) {
        this.client = client;
    }
    MonitoringSettings.prototype.get = function (resourceGroupName, serviceName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            options: options
        }, getOperationSpec, callback);
    };
    /**
     * Update the Monitoring Setting.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serviceName The name of the Service resource.
     * @param monitoringSettingResource Parameters for the update operation
     * @param [options] The optional parameters
     * @returns Promise<Models.MonitoringSettingsUpdatePutResponse>
     */
    MonitoringSettings.prototype.updatePut = function (resourceGroupName, serviceName, monitoringSettingResource, options) {
        return this.beginUpdatePut(resourceGroupName, serviceName, monitoringSettingResource, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Update the Monitoring Setting.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serviceName The name of the Service resource.
     * @param monitoringSettingResource Parameters for the update operation
     * @param [options] The optional parameters
     * @returns Promise<Models.MonitoringSettingsUpdatePatchResponse>
     */
    MonitoringSettings.prototype.updatePatch = function (resourceGroupName, serviceName, monitoringSettingResource, options) {
        return this.beginUpdatePatch(resourceGroupName, serviceName, monitoringSettingResource, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Update the Monitoring Setting.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serviceName The name of the Service resource.
     * @param monitoringSettingResource Parameters for the update operation
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    MonitoringSettings.prototype.beginUpdatePut = function (resourceGroupName, serviceName, monitoringSettingResource, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            monitoringSettingResource: monitoringSettingResource,
            options: options
        }, beginUpdatePutOperationSpec, options);
    };
    /**
     * Update the Monitoring Setting.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serviceName The name of the Service resource.
     * @param monitoringSettingResource Parameters for the update operation
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    MonitoringSettings.prototype.beginUpdatePatch = function (resourceGroupName, serviceName, monitoringSettingResource, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            monitoringSettingResource: monitoringSettingResource,
            options: options
        }, beginUpdatePatchOperationSpec, options);
    };
    return MonitoringSettings;
}());
export { MonitoringSettings };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AppPlatform/Spring/{serviceName}/monitoringSettings/default",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serviceName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.MonitoringSettingResource
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginUpdatePutOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AppPlatform/Spring/{serviceName}/monitoringSettings/default",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serviceName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "monitoringSettingResource",
        mapper: __assign(__assign({}, Mappers.MonitoringSettingResource), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.MonitoringSettingResource
        },
        202: {
            bodyMapper: Mappers.MonitoringSettingResource
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginUpdatePatchOperationSpec = {
    httpMethod: "PATCH",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AppPlatform/Spring/{serviceName}/monitoringSettings/default",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serviceName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "monitoringSettingResource",
        mapper: __assign(__assign({}, Mappers.MonitoringSettingResource), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.MonitoringSettingResource
        },
        202: {
            bodyMapper: Mappers.MonitoringSettingResource
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
