// @azure/arm-appplatform/esm/operations/runtimeVersions.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/runtimeVersionsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a RuntimeVersions. */
var RuntimeVersions = /** @class */ (function () {
    /**
     * Create a RuntimeVersions.
     * @param {AppPlatformManagementClientContext} client Reference to the service client.
     */
    function RuntimeVersions(client) {
        this.client = client;
    }
    RuntimeVersions.prototype.listRuntimeVersions = function (options, callback) {
        return this.client.sendOperationRequest({
            options: options
        }, listRuntimeVersionsOperationSpec, callback);
    };
    return RuntimeVersions;
}());
export { RuntimeVersions };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listRuntimeVersionsOperationSpec = {
    httpMethod: "GET",
    path: "providers/Microsoft.AppPlatform/runtimeVersions",
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.AvailableRuntimeVersions
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
