// @azure/arm-appplatform/dist-esm/src/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link ProvisioningState} that the service accepts. */
export var KnownProvisioningState;
(function (KnownProvisioningState) {
    /** Creating */
    KnownProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownProvisioningState["Updating"] = "Updating";
    /** Starting */
    KnownProvisioningState["Starting"] = "Starting";
    /** Stopping */
    KnownProvisioningState["Stopping"] = "Stopping";
    /** Deleting */
    KnownProvisioningState["Deleting"] = "Deleting";
    /** Deleted */
    KnownProvisioningState["Deleted"] = "Deleted";
    /** Succeeded */
    KnownProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownProvisioningState["Failed"] = "Failed";
    /** Moving */
    KnownProvisioningState["Moving"] = "Moving";
    /** Moved */
    KnownProvisioningState["Moved"] = "Moved";
    /** MoveFailed */
    KnownProvisioningState["MoveFailed"] = "MoveFailed";
})(KnownProvisioningState || (KnownProvisioningState = {}));
/** Known values of {@link TrafficDirection} that the service accepts. */
export var KnownTrafficDirection;
(function (KnownTrafficDirection) {
    /** Inbound */
    KnownTrafficDirection["Inbound"] = "Inbound";
    /** Outbound */
    KnownTrafficDirection["Outbound"] = "Outbound";
})(KnownTrafficDirection || (KnownTrafficDirection = {}));
/** Known values of {@link PowerState} that the service accepts. */
export var KnownPowerState;
(function (KnownPowerState) {
    /** Running */
    KnownPowerState["Running"] = "Running";
    /** Stopped */
    KnownPowerState["Stopped"] = "Stopped";
})(KnownPowerState || (KnownPowerState = {}));
/** Known values of {@link CreatedByType} that the service accepts. */
export var KnownCreatedByType;
(function (KnownCreatedByType) {
    /** User */
    KnownCreatedByType["User"] = "User";
    /** Application */
    KnownCreatedByType["Application"] = "Application";
    /** ManagedIdentity */
    KnownCreatedByType["ManagedIdentity"] = "ManagedIdentity";
    /** Key */
    KnownCreatedByType["Key"] = "Key";
})(KnownCreatedByType || (KnownCreatedByType = {}));
/** Known values of {@link LastModifiedByType} that the service accepts. */
export var KnownLastModifiedByType;
(function (KnownLastModifiedByType) {
    /** User */
    KnownLastModifiedByType["User"] = "User";
    /** Application */
    KnownLastModifiedByType["Application"] = "Application";
    /** ManagedIdentity */
    KnownLastModifiedByType["ManagedIdentity"] = "ManagedIdentity";
    /** Key */
    KnownLastModifiedByType["Key"] = "Key";
})(KnownLastModifiedByType || (KnownLastModifiedByType = {}));
/** Known values of {@link TestKeyType} that the service accepts. */
export var KnownTestKeyType;
(function (KnownTestKeyType) {
    /** Primary */
    KnownTestKeyType["Primary"] = "Primary";
    /** Secondary */
    KnownTestKeyType["Secondary"] = "Secondary";
})(KnownTestKeyType || (KnownTestKeyType = {}));
/** Known values of {@link ConfigServerState} that the service accepts. */
export var KnownConfigServerState;
(function (KnownConfigServerState) {
    /** NotAvailable */
    KnownConfigServerState["NotAvailable"] = "NotAvailable";
    /** Deleted */
    KnownConfigServerState["Deleted"] = "Deleted";
    /** Failed */
    KnownConfigServerState["Failed"] = "Failed";
    /** Succeeded */
    KnownConfigServerState["Succeeded"] = "Succeeded";
    /** Updating */
    KnownConfigServerState["Updating"] = "Updating";
})(KnownConfigServerState || (KnownConfigServerState = {}));
/** Known values of {@link ConfigurationServiceProvisioningState} that the service accepts. */
export var KnownConfigurationServiceProvisioningState;
(function (KnownConfigurationServiceProvisioningState) {
    /** Creating */
    KnownConfigurationServiceProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownConfigurationServiceProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownConfigurationServiceProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownConfigurationServiceProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownConfigurationServiceProvisioningState["Deleting"] = "Deleting";
})(KnownConfigurationServiceProvisioningState || (KnownConfigurationServiceProvisioningState = {}));
/** Known values of {@link ServiceRegistryProvisioningState} that the service accepts. */
export var KnownServiceRegistryProvisioningState;
(function (KnownServiceRegistryProvisioningState) {
    /** Creating */
    KnownServiceRegistryProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownServiceRegistryProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownServiceRegistryProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownServiceRegistryProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownServiceRegistryProvisioningState["Deleting"] = "Deleting";
})(KnownServiceRegistryProvisioningState || (KnownServiceRegistryProvisioningState = {}));
/** Known values of {@link ApplicationLiveViewProvisioningState} that the service accepts. */
export var KnownApplicationLiveViewProvisioningState;
(function (KnownApplicationLiveViewProvisioningState) {
    /** Creating */
    KnownApplicationLiveViewProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownApplicationLiveViewProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownApplicationLiveViewProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownApplicationLiveViewProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownApplicationLiveViewProvisioningState["Deleting"] = "Deleting";
    /** Canceled */
    KnownApplicationLiveViewProvisioningState["Canceled"] = "Canceled";
})(KnownApplicationLiveViewProvisioningState || (KnownApplicationLiveViewProvisioningState = {}));
/** Known values of {@link DevToolPortalProvisioningState} that the service accepts. */
export var KnownDevToolPortalProvisioningState;
(function (KnownDevToolPortalProvisioningState) {
    /** Creating */
    KnownDevToolPortalProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownDevToolPortalProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownDevToolPortalProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownDevToolPortalProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownDevToolPortalProvisioningState["Deleting"] = "Deleting";
    /** Canceled */
    KnownDevToolPortalProvisioningState["Canceled"] = "Canceled";
})(KnownDevToolPortalProvisioningState || (KnownDevToolPortalProvisioningState = {}));
/** Known values of {@link DevToolPortalFeatureState} that the service accepts. */
export var KnownDevToolPortalFeatureState;
(function (KnownDevToolPortalFeatureState) {
    /** Enable the plugin in Dev Tool Portal. */
    KnownDevToolPortalFeatureState["Enabled"] = "Enabled";
    /** Disable the plugin in Dev Tool Portal. */
    KnownDevToolPortalFeatureState["Disabled"] = "Disabled";
})(KnownDevToolPortalFeatureState || (KnownDevToolPortalFeatureState = {}));
/** Known values of {@link ContainerRegistryProvisioningState} that the service accepts. */
export var KnownContainerRegistryProvisioningState;
(function (KnownContainerRegistryProvisioningState) {
    /** Creating */
    KnownContainerRegistryProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownContainerRegistryProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownContainerRegistryProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownContainerRegistryProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownContainerRegistryProvisioningState["Canceled"] = "Canceled";
})(KnownContainerRegistryProvisioningState || (KnownContainerRegistryProvisioningState = {}));
/** Known values of {@link BuildServiceProvisioningState} that the service accepts. */
export var KnownBuildServiceProvisioningState;
(function (KnownBuildServiceProvisioningState) {
    /** Creating */
    KnownBuildServiceProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownBuildServiceProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownBuildServiceProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownBuildServiceProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownBuildServiceProvisioningState["Deleting"] = "Deleting";
})(KnownBuildServiceProvisioningState || (KnownBuildServiceProvisioningState = {}));
/** Known values of {@link BuildProvisioningState} that the service accepts. */
export var KnownBuildProvisioningState;
(function (KnownBuildProvisioningState) {
    /** Creating */
    KnownBuildProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownBuildProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownBuildProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownBuildProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownBuildProvisioningState["Deleting"] = "Deleting";
})(KnownBuildProvisioningState || (KnownBuildProvisioningState = {}));
/** Known values of {@link BindingType} that the service accepts. */
export var KnownBindingType;
(function (KnownBindingType) {
    /** ApplicationInsights */
    KnownBindingType["ApplicationInsights"] = "ApplicationInsights";
    /** ApacheSkyWalking */
    KnownBindingType["ApacheSkyWalking"] = "ApacheSkyWalking";
    /** AppDynamics */
    KnownBindingType["AppDynamics"] = "AppDynamics";
    /** Dynatrace */
    KnownBindingType["Dynatrace"] = "Dynatrace";
    /** NewRelic */
    KnownBindingType["NewRelic"] = "NewRelic";
    /** ElasticAPM */
    KnownBindingType["ElasticAPM"] = "ElasticAPM";
    /** CACertificates */
    KnownBindingType["CACertificates"] = "CACertificates";
})(KnownBindingType || (KnownBindingType = {}));
/** Known values of {@link BuildpackBindingProvisioningState} that the service accepts. */
export var KnownBuildpackBindingProvisioningState;
(function (KnownBuildpackBindingProvisioningState) {
    /** Creating */
    KnownBuildpackBindingProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownBuildpackBindingProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownBuildpackBindingProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownBuildpackBindingProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownBuildpackBindingProvisioningState["Deleting"] = "Deleting";
})(KnownBuildpackBindingProvisioningState || (KnownBuildpackBindingProvisioningState = {}));
/** Known values of {@link BuildResultProvisioningState} that the service accepts. */
export var KnownBuildResultProvisioningState;
(function (KnownBuildResultProvisioningState) {
    /** Queuing */
    KnownBuildResultProvisioningState["Queuing"] = "Queuing";
    /** Building */
    KnownBuildResultProvisioningState["Building"] = "Building";
    /** Succeeded */
    KnownBuildResultProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownBuildResultProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownBuildResultProvisioningState["Deleting"] = "Deleting";
})(KnownBuildResultProvisioningState || (KnownBuildResultProvisioningState = {}));
/** Known values of {@link KPackBuildStageProvisioningState} that the service accepts. */
export var KnownKPackBuildStageProvisioningState;
(function (KnownKPackBuildStageProvisioningState) {
    /** NotStarted */
    KnownKPackBuildStageProvisioningState["NotStarted"] = "NotStarted";
    /** Running */
    KnownKPackBuildStageProvisioningState["Running"] = "Running";
    /** Succeeded */
    KnownKPackBuildStageProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownKPackBuildStageProvisioningState["Failed"] = "Failed";
})(KnownKPackBuildStageProvisioningState || (KnownKPackBuildStageProvisioningState = {}));
/** Known values of {@link BuilderProvisioningState} that the service accepts. */
export var KnownBuilderProvisioningState;
(function (KnownBuilderProvisioningState) {
    /** Creating */
    KnownBuilderProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownBuilderProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownBuilderProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownBuilderProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownBuilderProvisioningState["Deleting"] = "Deleting";
})(KnownBuilderProvisioningState || (KnownBuilderProvisioningState = {}));
/** Known values of {@link MonitoringSettingState} that the service accepts. */
export var KnownMonitoringSettingState;
(function (KnownMonitoringSettingState) {
    /** NotAvailable */
    KnownMonitoringSettingState["NotAvailable"] = "NotAvailable";
    /** Failed */
    KnownMonitoringSettingState["Failed"] = "Failed";
    /** Succeeded */
    KnownMonitoringSettingState["Succeeded"] = "Succeeded";
    /** Updating */
    KnownMonitoringSettingState["Updating"] = "Updating";
})(KnownMonitoringSettingState || (KnownMonitoringSettingState = {}));
/** Known values of {@link AppResourceProvisioningState} that the service accepts. */
export var KnownAppResourceProvisioningState;
(function (KnownAppResourceProvisioningState) {
    /** Succeeded */
    KnownAppResourceProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownAppResourceProvisioningState["Failed"] = "Failed";
    /** Creating */
    KnownAppResourceProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownAppResourceProvisioningState["Updating"] = "Updating";
    /** Deleting */
    KnownAppResourceProvisioningState["Deleting"] = "Deleting";
})(KnownAppResourceProvisioningState || (KnownAppResourceProvisioningState = {}));
/** Known values of {@link Type} that the service accepts. */
export var KnownType;
(function (KnownType) {
    /** AzureFileVolume */
    KnownType["AzureFileVolume"] = "AzureFileVolume";
})(KnownType || (KnownType = {}));
/** Known values of {@link SessionAffinity} that the service accepts. */
export var KnownSessionAffinity;
(function (KnownSessionAffinity) {
    /** Cookie */
    KnownSessionAffinity["Cookie"] = "Cookie";
    /** None */
    KnownSessionAffinity["None"] = "None";
})(KnownSessionAffinity || (KnownSessionAffinity = {}));
/** Known values of {@link BackendProtocol} that the service accepts. */
export var KnownBackendProtocol;
(function (KnownBackendProtocol) {
    /** Grpc */
    KnownBackendProtocol["Grpc"] = "GRPC";
    /** Default */
    KnownBackendProtocol["Default"] = "Default";
})(KnownBackendProtocol || (KnownBackendProtocol = {}));
/** Known values of {@link ManagedIdentityType} that the service accepts. */
export var KnownManagedIdentityType;
(function (KnownManagedIdentityType) {
    /** None */
    KnownManagedIdentityType["None"] = "None";
    /** SystemAssigned */
    KnownManagedIdentityType["SystemAssigned"] = "SystemAssigned";
    /** UserAssigned */
    KnownManagedIdentityType["UserAssigned"] = "UserAssigned";
    /** SystemAssignedUserAssigned */
    KnownManagedIdentityType["SystemAssignedUserAssigned"] = "SystemAssigned,UserAssigned";
})(KnownManagedIdentityType || (KnownManagedIdentityType = {}));
/** Known values of {@link StorageType} that the service accepts. */
export var KnownStorageType;
(function (KnownStorageType) {
    /** StorageAccount */
    KnownStorageType["StorageAccount"] = "StorageAccount";
})(KnownStorageType || (KnownStorageType = {}));
/** Known values of {@link CertificateResourceProvisioningState} that the service accepts. */
export var KnownCertificateResourceProvisioningState;
(function (KnownCertificateResourceProvisioningState) {
    /** Creating */
    KnownCertificateResourceProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownCertificateResourceProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownCertificateResourceProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownCertificateResourceProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownCertificateResourceProvisioningState["Deleting"] = "Deleting";
})(KnownCertificateResourceProvisioningState || (KnownCertificateResourceProvisioningState = {}));
/** Known values of {@link CustomDomainResourceProvisioningState} that the service accepts. */
export var KnownCustomDomainResourceProvisioningState;
(function (KnownCustomDomainResourceProvisioningState) {
    /** Creating */
    KnownCustomDomainResourceProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownCustomDomainResourceProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownCustomDomainResourceProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownCustomDomainResourceProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownCustomDomainResourceProvisioningState["Deleting"] = "Deleting";
})(KnownCustomDomainResourceProvisioningState || (KnownCustomDomainResourceProvisioningState = {}));
/** Known values of {@link ProbeActionType} that the service accepts. */
export var KnownProbeActionType;
(function (KnownProbeActionType) {
    /** HttpGetAction */
    KnownProbeActionType["HttpGetAction"] = "HTTPGetAction";
    /** TCPSocketAction */
    KnownProbeActionType["TCPSocketAction"] = "TCPSocketAction";
    /** ExecAction */
    KnownProbeActionType["ExecAction"] = "ExecAction";
})(KnownProbeActionType || (KnownProbeActionType = {}));
/** Known values of {@link DeploymentResourceProvisioningState} that the service accepts. */
export var KnownDeploymentResourceProvisioningState;
(function (KnownDeploymentResourceProvisioningState) {
    /** Creating */
    KnownDeploymentResourceProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownDeploymentResourceProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownDeploymentResourceProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownDeploymentResourceProvisioningState["Failed"] = "Failed";
})(KnownDeploymentResourceProvisioningState || (KnownDeploymentResourceProvisioningState = {}));
/** Known values of {@link DeploymentResourceStatus} that the service accepts. */
export var KnownDeploymentResourceStatus;
(function (KnownDeploymentResourceStatus) {
    /** Stopped */
    KnownDeploymentResourceStatus["Stopped"] = "Stopped";
    /** Running */
    KnownDeploymentResourceStatus["Running"] = "Running";
})(KnownDeploymentResourceStatus || (KnownDeploymentResourceStatus = {}));
/** Known values of {@link ActionType} that the service accepts. */
export var KnownActionType;
(function (KnownActionType) {
    /** Internal */
    KnownActionType["Internal"] = "Internal";
})(KnownActionType || (KnownActionType = {}));
/** Known values of {@link SupportedRuntimeValue} that the service accepts. */
export var KnownSupportedRuntimeValue;
(function (KnownSupportedRuntimeValue) {
    /** Java8 */
    KnownSupportedRuntimeValue["Java8"] = "Java_8";
    /** Java11 */
    KnownSupportedRuntimeValue["Java11"] = "Java_11";
    /** Java17 */
    KnownSupportedRuntimeValue["Java17"] = "Java_17";
    /** NetCore31 */
    KnownSupportedRuntimeValue["NetCore31"] = "NetCore_31";
})(KnownSupportedRuntimeValue || (KnownSupportedRuntimeValue = {}));
/** Known values of {@link SupportedRuntimePlatform} that the service accepts. */
export var KnownSupportedRuntimePlatform;
(function (KnownSupportedRuntimePlatform) {
    /** Java */
    KnownSupportedRuntimePlatform["Java"] = "Java";
    /** NETCore */
    KnownSupportedRuntimePlatform["NETCore"] = ".NET Core";
})(KnownSupportedRuntimePlatform || (KnownSupportedRuntimePlatform = {}));
/** Known values of {@link SkuScaleType} that the service accepts. */
export var KnownSkuScaleType;
(function (KnownSkuScaleType) {
    /** None */
    KnownSkuScaleType["None"] = "None";
    /** Manual */
    KnownSkuScaleType["Manual"] = "Manual";
    /** Automatic */
    KnownSkuScaleType["Automatic"] = "Automatic";
})(KnownSkuScaleType || (KnownSkuScaleType = {}));
/** Known values of {@link ResourceSkuRestrictionsType} that the service accepts. */
export var KnownResourceSkuRestrictionsType;
(function (KnownResourceSkuRestrictionsType) {
    /** Location */
    KnownResourceSkuRestrictionsType["Location"] = "Location";
    /** Zone */
    KnownResourceSkuRestrictionsType["Zone"] = "Zone";
})(KnownResourceSkuRestrictionsType || (KnownResourceSkuRestrictionsType = {}));
/** Known values of {@link ResourceSkuRestrictionsReasonCode} that the service accepts. */
export var KnownResourceSkuRestrictionsReasonCode;
(function (KnownResourceSkuRestrictionsReasonCode) {
    /** QuotaId */
    KnownResourceSkuRestrictionsReasonCode["QuotaId"] = "QuotaId";
    /** NotAvailableForSubscription */
    KnownResourceSkuRestrictionsReasonCode["NotAvailableForSubscription"] = "NotAvailableForSubscription";
})(KnownResourceSkuRestrictionsReasonCode || (KnownResourceSkuRestrictionsReasonCode = {}));
/** Known values of {@link GatewayProvisioningState} that the service accepts. */
export var KnownGatewayProvisioningState;
(function (KnownGatewayProvisioningState) {
    /** Creating */
    KnownGatewayProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownGatewayProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownGatewayProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownGatewayProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownGatewayProvisioningState["Deleting"] = "Deleting";
})(KnownGatewayProvisioningState || (KnownGatewayProvisioningState = {}));
/** Known values of {@link GatewayCertificateVerification} that the service accepts. */
export var KnownGatewayCertificateVerification;
(function (KnownGatewayCertificateVerification) {
    /** Enable certificate verification in Spring Cloud Gateway. */
    KnownGatewayCertificateVerification["Enabled"] = "Enabled";
    /** Disable certificate verification in Spring Cloud Gateway. */
    KnownGatewayCertificateVerification["Disabled"] = "Disabled";
})(KnownGatewayCertificateVerification || (KnownGatewayCertificateVerification = {}));
/** Known values of {@link ApmType} that the service accepts. */
export var KnownApmType;
(function (KnownApmType) {
    /** ApplicationInsights */
    KnownApmType["ApplicationInsights"] = "ApplicationInsights";
    /** AppDynamics */
    KnownApmType["AppDynamics"] = "AppDynamics";
    /** Dynatrace */
    KnownApmType["Dynatrace"] = "Dynatrace";
    /** NewRelic */
    KnownApmType["NewRelic"] = "NewRelic";
    /** ElasticAPM */
    KnownApmType["ElasticAPM"] = "ElasticAPM";
})(KnownApmType || (KnownApmType = {}));
/** Known values of {@link GatewayRouteConfigProtocol} that the service accepts. */
export var KnownGatewayRouteConfigProtocol;
(function (KnownGatewayRouteConfigProtocol) {
    /** Http */
    KnownGatewayRouteConfigProtocol["Http"] = "HTTP";
    /** Https */
    KnownGatewayRouteConfigProtocol["Https"] = "HTTPS";
})(KnownGatewayRouteConfigProtocol || (KnownGatewayRouteConfigProtocol = {}));
/** Known values of {@link ApiPortalProvisioningState} that the service accepts. */
export var KnownApiPortalProvisioningState;
(function (KnownApiPortalProvisioningState) {
    /** Creating */
    KnownApiPortalProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownApiPortalProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownApiPortalProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownApiPortalProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownApiPortalProvisioningState["Deleting"] = "Deleting";
})(KnownApiPortalProvisioningState || (KnownApiPortalProvisioningState = {}));
/** Known values of {@link ApplicationAcceleratorProvisioningState} that the service accepts. */
export var KnownApplicationAcceleratorProvisioningState;
(function (KnownApplicationAcceleratorProvisioningState) {
    /** Creating */
    KnownApplicationAcceleratorProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownApplicationAcceleratorProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownApplicationAcceleratorProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownApplicationAcceleratorProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownApplicationAcceleratorProvisioningState["Deleting"] = "Deleting";
})(KnownApplicationAcceleratorProvisioningState || (KnownApplicationAcceleratorProvisioningState = {}));
/** Known values of {@link CustomizedAcceleratorProvisioningState} that the service accepts. */
export var KnownCustomizedAcceleratorProvisioningState;
(function (KnownCustomizedAcceleratorProvisioningState) {
    /** Creating */
    KnownCustomizedAcceleratorProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownCustomizedAcceleratorProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownCustomizedAcceleratorProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownCustomizedAcceleratorProvisioningState["Failed"] = "Failed";
    /** Deleting */
    KnownCustomizedAcceleratorProvisioningState["Deleting"] = "Deleting";
})(KnownCustomizedAcceleratorProvisioningState || (KnownCustomizedAcceleratorProvisioningState = {}));
/** Known values of {@link CustomizedAcceleratorValidateResultState} that the service accepts. */
export var KnownCustomizedAcceleratorValidateResultState;
(function (KnownCustomizedAcceleratorValidateResultState) {
    /** Customized accelerator properties are valid. */
    KnownCustomizedAcceleratorValidateResultState["Valid"] = "Valid";
    /** Customized accelerator properties are invalid. */
    KnownCustomizedAcceleratorValidateResultState["Invalid"] = "Invalid";
})(KnownCustomizedAcceleratorValidateResultState || (KnownCustomizedAcceleratorValidateResultState = {}));
/** Known values of {@link PredefinedAcceleratorProvisioningState} that the service accepts. */
export var KnownPredefinedAcceleratorProvisioningState;
(function (KnownPredefinedAcceleratorProvisioningState) {
    /** Creating */
    KnownPredefinedAcceleratorProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownPredefinedAcceleratorProvisioningState["Updating"] = "Updating";
    /** Succeeded */
    KnownPredefinedAcceleratorProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownPredefinedAcceleratorProvisioningState["Failed"] = "Failed";
})(KnownPredefinedAcceleratorProvisioningState || (KnownPredefinedAcceleratorProvisioningState = {}));
/** Known values of {@link PredefinedAcceleratorState} that the service accepts. */
export var KnownPredefinedAcceleratorState;
(function (KnownPredefinedAcceleratorState) {
    /** Enable the predefined accelerator. */
    KnownPredefinedAcceleratorState["Enabled"] = "Enabled";
    /** Disable the predefined accelerator. */
    KnownPredefinedAcceleratorState["Disabled"] = "Disabled";
})(KnownPredefinedAcceleratorState || (KnownPredefinedAcceleratorState = {}));
/** Known values of {@link HttpSchemeType} that the service accepts. */
export var KnownHttpSchemeType;
(function (KnownHttpSchemeType) {
    /** Http */
    KnownHttpSchemeType["Http"] = "HTTP";
    /** Https */
    KnownHttpSchemeType["Https"] = "HTTPS";
})(KnownHttpSchemeType || (KnownHttpSchemeType = {}));
