// @azure/arm-appplatform/dist-esm/src/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link ProvisioningState} that the service accepts. */
export var KnownProvisioningState;
(function (KnownProvisioningState) {
    KnownProvisioningState["Creating"] = "Creating";
    KnownProvisioningState["Updating"] = "Updating";
    KnownProvisioningState["Starting"] = "Starting";
    KnownProvisioningState["Stopping"] = "Stopping";
    KnownProvisioningState["Deleting"] = "Deleting";
    KnownProvisioningState["Deleted"] = "Deleted";
    KnownProvisioningState["Succeeded"] = "Succeeded";
    KnownProvisioningState["Failed"] = "Failed";
    KnownProvisioningState["Moving"] = "Moving";
    KnownProvisioningState["Moved"] = "Moved";
    KnownProvisioningState["MoveFailed"] = "MoveFailed";
})(KnownProvisioningState || (KnownProvisioningState = {}));
/** Known values of {@link TrafficDirection} that the service accepts. */
export var KnownTrafficDirection;
(function (KnownTrafficDirection) {
    KnownTrafficDirection["Inbound"] = "Inbound";
    KnownTrafficDirection["Outbound"] = "Outbound";
})(KnownTrafficDirection || (KnownTrafficDirection = {}));
/** Known values of {@link CreatedByType} that the service accepts. */
export var KnownCreatedByType;
(function (KnownCreatedByType) {
    KnownCreatedByType["User"] = "User";
    KnownCreatedByType["Application"] = "Application";
    KnownCreatedByType["ManagedIdentity"] = "ManagedIdentity";
    KnownCreatedByType["Key"] = "Key";
})(KnownCreatedByType || (KnownCreatedByType = {}));
/** Known values of {@link LastModifiedByType} that the service accepts. */
export var KnownLastModifiedByType;
(function (KnownLastModifiedByType) {
    KnownLastModifiedByType["User"] = "User";
    KnownLastModifiedByType["Application"] = "Application";
    KnownLastModifiedByType["ManagedIdentity"] = "ManagedIdentity";
    KnownLastModifiedByType["Key"] = "Key";
})(KnownLastModifiedByType || (KnownLastModifiedByType = {}));
/** Known values of {@link TestKeyType} that the service accepts. */
export var KnownTestKeyType;
(function (KnownTestKeyType) {
    KnownTestKeyType["Primary"] = "Primary";
    KnownTestKeyType["Secondary"] = "Secondary";
})(KnownTestKeyType || (KnownTestKeyType = {}));
/** Known values of {@link ConfigServerState} that the service accepts. */
export var KnownConfigServerState;
(function (KnownConfigServerState) {
    KnownConfigServerState["NotAvailable"] = "NotAvailable";
    KnownConfigServerState["Deleted"] = "Deleted";
    KnownConfigServerState["Failed"] = "Failed";
    KnownConfigServerState["Succeeded"] = "Succeeded";
    KnownConfigServerState["Updating"] = "Updating";
})(KnownConfigServerState || (KnownConfigServerState = {}));
/** Known values of {@link ConfigurationServiceProvisioningState} that the service accepts. */
export var KnownConfigurationServiceProvisioningState;
(function (KnownConfigurationServiceProvisioningState) {
    KnownConfigurationServiceProvisioningState["Creating"] = "Creating";
    KnownConfigurationServiceProvisioningState["Updating"] = "Updating";
    KnownConfigurationServiceProvisioningState["Succeeded"] = "Succeeded";
    KnownConfigurationServiceProvisioningState["Failed"] = "Failed";
    KnownConfigurationServiceProvisioningState["Deleting"] = "Deleting";
})(KnownConfigurationServiceProvisioningState || (KnownConfigurationServiceProvisioningState = {}));
/** Known values of {@link ServiceRegistryProvisioningState} that the service accepts. */
export var KnownServiceRegistryProvisioningState;
(function (KnownServiceRegistryProvisioningState) {
    KnownServiceRegistryProvisioningState["Creating"] = "Creating";
    KnownServiceRegistryProvisioningState["Updating"] = "Updating";
    KnownServiceRegistryProvisioningState["Succeeded"] = "Succeeded";
    KnownServiceRegistryProvisioningState["Failed"] = "Failed";
    KnownServiceRegistryProvisioningState["Deleting"] = "Deleting";
})(KnownServiceRegistryProvisioningState || (KnownServiceRegistryProvisioningState = {}));
/** Known values of {@link BuildServiceProvisioningState} that the service accepts. */
export var KnownBuildServiceProvisioningState;
(function (KnownBuildServiceProvisioningState) {
    KnownBuildServiceProvisioningState["Creating"] = "Creating";
    KnownBuildServiceProvisioningState["Updating"] = "Updating";
    KnownBuildServiceProvisioningState["Succeeded"] = "Succeeded";
    KnownBuildServiceProvisioningState["Failed"] = "Failed";
    KnownBuildServiceProvisioningState["Deleting"] = "Deleting";
})(KnownBuildServiceProvisioningState || (KnownBuildServiceProvisioningState = {}));
/** Known values of {@link BuildProvisioningState} that the service accepts. */
export var KnownBuildProvisioningState;
(function (KnownBuildProvisioningState) {
    KnownBuildProvisioningState["Creating"] = "Creating";
    KnownBuildProvisioningState["Updating"] = "Updating";
    KnownBuildProvisioningState["Succeeded"] = "Succeeded";
    KnownBuildProvisioningState["Failed"] = "Failed";
    KnownBuildProvisioningState["Deleting"] = "Deleting";
})(KnownBuildProvisioningState || (KnownBuildProvisioningState = {}));
/** Known values of {@link BindingType} that the service accepts. */
export var KnownBindingType;
(function (KnownBindingType) {
    KnownBindingType["ApplicationInsights"] = "ApplicationInsights";
    KnownBindingType["ApacheSkyWalking"] = "ApacheSkyWalking";
    KnownBindingType["AppDynamics"] = "AppDynamics";
    KnownBindingType["Dynatrace"] = "Dynatrace";
    KnownBindingType["NewRelic"] = "NewRelic";
    KnownBindingType["ElasticAPM"] = "ElasticAPM";
})(KnownBindingType || (KnownBindingType = {}));
/** Known values of {@link BuildpackBindingProvisioningState} that the service accepts. */
export var KnownBuildpackBindingProvisioningState;
(function (KnownBuildpackBindingProvisioningState) {
    KnownBuildpackBindingProvisioningState["Creating"] = "Creating";
    KnownBuildpackBindingProvisioningState["Updating"] = "Updating";
    KnownBuildpackBindingProvisioningState["Succeeded"] = "Succeeded";
    KnownBuildpackBindingProvisioningState["Failed"] = "Failed";
    KnownBuildpackBindingProvisioningState["Deleting"] = "Deleting";
})(KnownBuildpackBindingProvisioningState || (KnownBuildpackBindingProvisioningState = {}));
/** Known values of {@link BuildResultProvisioningState} that the service accepts. */
export var KnownBuildResultProvisioningState;
(function (KnownBuildResultProvisioningState) {
    KnownBuildResultProvisioningState["Queuing"] = "Queuing";
    KnownBuildResultProvisioningState["Building"] = "Building";
    KnownBuildResultProvisioningState["Succeeded"] = "Succeeded";
    KnownBuildResultProvisioningState["Failed"] = "Failed";
    KnownBuildResultProvisioningState["Deleting"] = "Deleting";
})(KnownBuildResultProvisioningState || (KnownBuildResultProvisioningState = {}));
/** Known values of {@link KPackBuildStageProvisioningState} that the service accepts. */
export var KnownKPackBuildStageProvisioningState;
(function (KnownKPackBuildStageProvisioningState) {
    KnownKPackBuildStageProvisioningState["NotStarted"] = "NotStarted";
    KnownKPackBuildStageProvisioningState["Running"] = "Running";
    KnownKPackBuildStageProvisioningState["Succeeded"] = "Succeeded";
    KnownKPackBuildStageProvisioningState["Failed"] = "Failed";
})(KnownKPackBuildStageProvisioningState || (KnownKPackBuildStageProvisioningState = {}));
/** Known values of {@link BuilderProvisioningState} that the service accepts. */
export var KnownBuilderProvisioningState;
(function (KnownBuilderProvisioningState) {
    KnownBuilderProvisioningState["Creating"] = "Creating";
    KnownBuilderProvisioningState["Updating"] = "Updating";
    KnownBuilderProvisioningState["Succeeded"] = "Succeeded";
    KnownBuilderProvisioningState["Failed"] = "Failed";
    KnownBuilderProvisioningState["Deleting"] = "Deleting";
})(KnownBuilderProvisioningState || (KnownBuilderProvisioningState = {}));
/** Known values of {@link MonitoringSettingState} that the service accepts. */
export var KnownMonitoringSettingState;
(function (KnownMonitoringSettingState) {
    KnownMonitoringSettingState["NotAvailable"] = "NotAvailable";
    KnownMonitoringSettingState["Failed"] = "Failed";
    KnownMonitoringSettingState["Succeeded"] = "Succeeded";
    KnownMonitoringSettingState["Updating"] = "Updating";
})(KnownMonitoringSettingState || (KnownMonitoringSettingState = {}));
/** Known values of {@link AppResourceProvisioningState} that the service accepts. */
export var KnownAppResourceProvisioningState;
(function (KnownAppResourceProvisioningState) {
    KnownAppResourceProvisioningState["Succeeded"] = "Succeeded";
    KnownAppResourceProvisioningState["Failed"] = "Failed";
    KnownAppResourceProvisioningState["Creating"] = "Creating";
    KnownAppResourceProvisioningState["Updating"] = "Updating";
    KnownAppResourceProvisioningState["Deleting"] = "Deleting";
})(KnownAppResourceProvisioningState || (KnownAppResourceProvisioningState = {}));
/** Known values of {@link ManagedIdentityType} that the service accepts. */
export var KnownManagedIdentityType;
(function (KnownManagedIdentityType) {
    KnownManagedIdentityType["None"] = "None";
    KnownManagedIdentityType["SystemAssigned"] = "SystemAssigned";
    KnownManagedIdentityType["UserAssigned"] = "UserAssigned";
    KnownManagedIdentityType["SystemAssignedUserAssigned"] = "SystemAssigned,UserAssigned";
})(KnownManagedIdentityType || (KnownManagedIdentityType = {}));
/** Known values of {@link DeploymentResourceProvisioningState} that the service accepts. */
export var KnownDeploymentResourceProvisioningState;
(function (KnownDeploymentResourceProvisioningState) {
    KnownDeploymentResourceProvisioningState["Creating"] = "Creating";
    KnownDeploymentResourceProvisioningState["Updating"] = "Updating";
    KnownDeploymentResourceProvisioningState["Succeeded"] = "Succeeded";
    KnownDeploymentResourceProvisioningState["Failed"] = "Failed";
})(KnownDeploymentResourceProvisioningState || (KnownDeploymentResourceProvisioningState = {}));
/** Known values of {@link DeploymentResourceStatus} that the service accepts. */
export var KnownDeploymentResourceStatus;
(function (KnownDeploymentResourceStatus) {
    KnownDeploymentResourceStatus["Stopped"] = "Stopped";
    KnownDeploymentResourceStatus["Running"] = "Running";
})(KnownDeploymentResourceStatus || (KnownDeploymentResourceStatus = {}));
/** Known values of {@link ActionType} that the service accepts. */
export var KnownActionType;
(function (KnownActionType) {
    KnownActionType["Internal"] = "Internal";
})(KnownActionType || (KnownActionType = {}));
/** Known values of {@link SupportedRuntimeValue} that the service accepts. */
export var KnownSupportedRuntimeValue;
(function (KnownSupportedRuntimeValue) {
    KnownSupportedRuntimeValue["Java8"] = "Java_8";
    KnownSupportedRuntimeValue["Java11"] = "Java_11";
    KnownSupportedRuntimeValue["Java17"] = "Java_17";
    KnownSupportedRuntimeValue["NetCore31"] = "NetCore_31";
})(KnownSupportedRuntimeValue || (KnownSupportedRuntimeValue = {}));
/** Known values of {@link SupportedRuntimePlatform} that the service accepts. */
export var KnownSupportedRuntimePlatform;
(function (KnownSupportedRuntimePlatform) {
    KnownSupportedRuntimePlatform["Java"] = "Java";
    KnownSupportedRuntimePlatform["NETCore"] = ".NET Core";
})(KnownSupportedRuntimePlatform || (KnownSupportedRuntimePlatform = {}));
/** Known values of {@link SkuScaleType} that the service accepts. */
export var KnownSkuScaleType;
(function (KnownSkuScaleType) {
    KnownSkuScaleType["None"] = "None";
    KnownSkuScaleType["Manual"] = "Manual";
    KnownSkuScaleType["Automatic"] = "Automatic";
})(KnownSkuScaleType || (KnownSkuScaleType = {}));
/** Known values of {@link ResourceSkuRestrictionsType} that the service accepts. */
export var KnownResourceSkuRestrictionsType;
(function (KnownResourceSkuRestrictionsType) {
    KnownResourceSkuRestrictionsType["Location"] = "Location";
    KnownResourceSkuRestrictionsType["Zone"] = "Zone";
})(KnownResourceSkuRestrictionsType || (KnownResourceSkuRestrictionsType = {}));
/** Known values of {@link ResourceSkuRestrictionsReasonCode} that the service accepts. */
export var KnownResourceSkuRestrictionsReasonCode;
(function (KnownResourceSkuRestrictionsReasonCode) {
    KnownResourceSkuRestrictionsReasonCode["QuotaId"] = "QuotaId";
    KnownResourceSkuRestrictionsReasonCode["NotAvailableForSubscription"] = "NotAvailableForSubscription";
})(KnownResourceSkuRestrictionsReasonCode || (KnownResourceSkuRestrictionsReasonCode = {}));
