// @azure/arm-appplatform/dist-esm/src/appPlatformManagementClient.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { __awaiter } from "tslib";
import * as coreClient from "@azure/core-client";
import * as coreRestPipeline from "@azure/core-rest-pipeline";
import { ServicesImpl, ConfigServersImpl, ConfigurationServicesImpl, ServiceRegistriesImpl, ApplicationLiveViewsImpl, DevToolPortalsImpl, ContainerRegistriesImpl, BuildServiceOperationsImpl, BuildpackBindingImpl, BuildServiceBuilderImpl, BuildServiceAgentPoolImpl, MonitoringSettingsImpl, AppsImpl, BindingsImpl, StoragesImpl, CertificatesImpl, CustomDomainsImpl, DeploymentsImpl, OperationsImpl, RuntimeVersionsImpl, SkusImpl, GatewaysImpl, GatewayRouteConfigsImpl, GatewayCustomDomainsImpl, ApiPortalsImpl, ApiPortalCustomDomainsImpl, ApplicationAcceleratorsImpl, CustomizedAcceleratorsImpl, PredefinedAcceleratorsImpl } from "./operations";
export class AppPlatformManagementClient extends coreClient.ServiceClient {
    /**
     * Initializes a new instance of the AppPlatformManagementClient class.
     * @param credentials Subscription credentials which uniquely identify client subscription.
     * @param subscriptionId Gets subscription ID which uniquely identify the Microsoft Azure subscription.
     *                       The subscription ID forms part of the URI for every service call.
     * @param options The parameter options
     */
    constructor(credentials, subscriptionId, options) {
        var _a, _b, _c;
        if (credentials === undefined) {
            throw new Error("'credentials' cannot be null");
        }
        if (subscriptionId === undefined) {
            throw new Error("'subscriptionId' cannot be null");
        }
        // Initializing default values for options
        if (!options) {
            options = {};
        }
        const defaults = {
            requestContentType: "application/json; charset=utf-8",
            credential: credentials
        };
        const packageDetails = `azsdk-js-arm-appplatform/3.0.0-beta.1`;
        const userAgentPrefix = options.userAgentOptions && options.userAgentOptions.userAgentPrefix
            ? `${options.userAgentOptions.userAgentPrefix} ${packageDetails}`
            : `${packageDetails}`;
        const optionsWithDefaults = Object.assign(Object.assign(Object.assign({}, defaults), options), { userAgentOptions: {
                userAgentPrefix
            }, endpoint: (_b = (_a = options.endpoint) !== null && _a !== void 0 ? _a : options.baseUri) !== null && _b !== void 0 ? _b : "https://management.azure.com" });
        super(optionsWithDefaults);
        let bearerTokenAuthenticationPolicyFound = false;
        if ((options === null || options === void 0 ? void 0 : options.pipeline) && options.pipeline.getOrderedPolicies().length > 0) {
            const pipelinePolicies = options.pipeline.getOrderedPolicies();
            bearerTokenAuthenticationPolicyFound = pipelinePolicies.some((pipelinePolicy) => pipelinePolicy.name ===
                coreRestPipeline.bearerTokenAuthenticationPolicyName);
        }
        if (!options ||
            !options.pipeline ||
            options.pipeline.getOrderedPolicies().length == 0 ||
            !bearerTokenAuthenticationPolicyFound) {
            this.pipeline.removePolicy({
                name: coreRestPipeline.bearerTokenAuthenticationPolicyName
            });
            this.pipeline.addPolicy(coreRestPipeline.bearerTokenAuthenticationPolicy({
                credential: credentials,
                scopes: (_c = optionsWithDefaults.credentialScopes) !== null && _c !== void 0 ? _c : `${optionsWithDefaults.endpoint}/.default`,
                challengeCallbacks: {
                    authorizeRequestOnChallenge: coreClient.authorizeRequestOnClaimChallenge
                }
            }));
        }
        // Parameter assignments
        this.subscriptionId = subscriptionId;
        // Assigning values to Constant parameters
        this.$host = options.$host || "https://management.azure.com";
        this.apiVersion = options.apiVersion || "2023-03-01-preview";
        this.services = new ServicesImpl(this);
        this.configServers = new ConfigServersImpl(this);
        this.configurationServices = new ConfigurationServicesImpl(this);
        this.serviceRegistries = new ServiceRegistriesImpl(this);
        this.applicationLiveViews = new ApplicationLiveViewsImpl(this);
        this.devToolPortals = new DevToolPortalsImpl(this);
        this.containerRegistries = new ContainerRegistriesImpl(this);
        this.buildServiceOperations = new BuildServiceOperationsImpl(this);
        this.buildpackBinding = new BuildpackBindingImpl(this);
        this.buildServiceBuilder = new BuildServiceBuilderImpl(this);
        this.buildServiceAgentPool = new BuildServiceAgentPoolImpl(this);
        this.monitoringSettings = new MonitoringSettingsImpl(this);
        this.apps = new AppsImpl(this);
        this.bindings = new BindingsImpl(this);
        this.storages = new StoragesImpl(this);
        this.certificates = new CertificatesImpl(this);
        this.customDomains = new CustomDomainsImpl(this);
        this.deployments = new DeploymentsImpl(this);
        this.operations = new OperationsImpl(this);
        this.runtimeVersions = new RuntimeVersionsImpl(this);
        this.skus = new SkusImpl(this);
        this.gateways = new GatewaysImpl(this);
        this.gatewayRouteConfigs = new GatewayRouteConfigsImpl(this);
        this.gatewayCustomDomains = new GatewayCustomDomainsImpl(this);
        this.apiPortals = new ApiPortalsImpl(this);
        this.apiPortalCustomDomains = new ApiPortalCustomDomainsImpl(this);
        this.applicationAccelerators = new ApplicationAcceleratorsImpl(this);
        this.customizedAccelerators = new CustomizedAcceleratorsImpl(this);
        this.predefinedAccelerators = new PredefinedAcceleratorsImpl(this);
        this.addCustomApiVersionPolicy(options.apiVersion);
    }
    /** A function that adds a policy that sets the api-version (or equivalent) to reflect the library version. */
    addCustomApiVersionPolicy(apiVersion) {
        if (!apiVersion) {
            return;
        }
        const apiVersionPolicy = {
            name: "CustomApiVersionPolicy",
            sendRequest(request, next) {
                return __awaiter(this, void 0, void 0, function* () {
                    const param = request.url.split("?");
                    if (param.length > 1) {
                        const newParams = param[1].split("&").map((item) => {
                            if (item.indexOf("api-version") > -1) {
                                return "api-version=" + apiVersion;
                            }
                            else {
                                return item;
                            }
                        });
                        request.url = param[0] + "?" + newParams.join("&");
                    }
                    return next(request);
                });
            }
        };
        this.pipeline.addPolicy(apiVersionPolicy);
    }
}
