// @azure/arm-appplatform/esm/models/mappers.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { __assign } from "tslib";
import { CloudErrorMapper, BaseResourceMapper } from "@azure/ms-rest-azure-js";
export var CloudError = CloudErrorMapper;
export var BaseResource = BaseResourceMapper;
export var NetworkProfileOutboundIPs = {
    serializedName: "NetworkProfile_outboundIPs",
    type: {
        name: "Composite",
        className: "NetworkProfileOutboundIPs",
        modelProperties: {
            publicIPs: {
                readOnly: true,
                serializedName: "publicIPs",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            }
        }
    }
};
export var RequiredTraffic = {
    serializedName: "RequiredTraffic",
    type: {
        name: "Composite",
        className: "RequiredTraffic",
        modelProperties: {
            protocol: {
                readOnly: true,
                serializedName: "protocol",
                type: {
                    name: "String"
                }
            },
            port: {
                readOnly: true,
                serializedName: "port",
                type: {
                    name: "Number"
                }
            },
            ips: {
                readOnly: true,
                serializedName: "ips",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            fqdns: {
                readOnly: true,
                serializedName: "fqdns",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            direction: {
                readOnly: true,
                serializedName: "direction",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var NetworkProfile = {
    serializedName: "NetworkProfile",
    type: {
        name: "Composite",
        className: "NetworkProfile",
        modelProperties: {
            serviceRuntimeSubnetId: {
                serializedName: "serviceRuntimeSubnetId",
                type: {
                    name: "String"
                }
            },
            appSubnetId: {
                serializedName: "appSubnetId",
                type: {
                    name: "String"
                }
            },
            serviceCidr: {
                serializedName: "serviceCidr",
                type: {
                    name: "String"
                }
            },
            serviceRuntimeNetworkResourceGroup: {
                serializedName: "serviceRuntimeNetworkResourceGroup",
                type: {
                    name: "String"
                }
            },
            appNetworkResourceGroup: {
                serializedName: "appNetworkResourceGroup",
                type: {
                    name: "String"
                }
            },
            outboundIPs: {
                readOnly: true,
                serializedName: "outboundIPs",
                type: {
                    name: "Composite",
                    className: "NetworkProfileOutboundIPs"
                }
            },
            requiredTraffics: {
                readOnly: true,
                serializedName: "requiredTraffics",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "RequiredTraffic"
                        }
                    }
                }
            }
        }
    }
};
export var ClusterResourceProperties = {
    serializedName: "ClusterResourceProperties",
    type: {
        name: "Composite",
        className: "ClusterResourceProperties",
        modelProperties: {
            provisioningState: {
                readOnly: true,
                serializedName: "provisioningState",
                type: {
                    name: "String"
                }
            },
            networkProfile: {
                serializedName: "networkProfile",
                type: {
                    name: "Composite",
                    className: "NetworkProfile"
                }
            },
            version: {
                readOnly: true,
                serializedName: "version",
                type: {
                    name: "Number"
                }
            },
            serviceId: {
                readOnly: true,
                serializedName: "serviceId",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var Sku = {
    serializedName: "Sku",
    type: {
        name: "Composite",
        className: "Sku",
        modelProperties: {
            name: {
                serializedName: "name",
                defaultValue: 'S0',
                type: {
                    name: "String"
                }
            },
            tier: {
                serializedName: "tier",
                defaultValue: 'Standard',
                type: {
                    name: "String"
                }
            },
            capacity: {
                serializedName: "capacity",
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var Resource = {
    serializedName: "Resource",
    type: {
        name: "Composite",
        className: "Resource",
        modelProperties: {
            id: {
                readOnly: true,
                serializedName: "id",
                type: {
                    name: "String"
                }
            },
            name: {
                readOnly: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            type: {
                readOnly: true,
                serializedName: "type",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var TrackedResource = {
    serializedName: "TrackedResource",
    type: {
        name: "Composite",
        className: "TrackedResource",
        modelProperties: __assign(__assign({}, Resource.type.modelProperties), { location: {
                serializedName: "location",
                type: {
                    name: "String"
                }
            }, tags: {
                serializedName: "tags",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                }
            } })
    }
};
export var ServiceResource = {
    serializedName: "ServiceResource",
    type: {
        name: "Composite",
        className: "ServiceResource",
        modelProperties: __assign(__assign({}, TrackedResource.type.modelProperties), { properties: {
                serializedName: "properties",
                type: {
                    name: "Composite",
                    className: "ClusterResourceProperties"
                }
            }, sku: {
                serializedName: "sku",
                type: {
                    name: "Composite",
                    className: "Sku"
                }
            } })
    }
};
export var ManagedIdentityProperties = {
    serializedName: "ManagedIdentityProperties",
    type: {
        name: "Composite",
        className: "ManagedIdentityProperties",
        modelProperties: {
            type: {
                serializedName: "type",
                type: {
                    name: "String"
                }
            },
            principalId: {
                serializedName: "principalId",
                type: {
                    name: "String"
                }
            },
            tenantId: {
                serializedName: "tenantId",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ConfigServerSettingsErrorRecord = {
    serializedName: "ConfigServerSettingsErrorRecord",
    type: {
        name: "Composite",
        className: "ConfigServerSettingsErrorRecord",
        modelProperties: {
            name: {
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            uri: {
                serializedName: "uri",
                type: {
                    name: "String"
                }
            },
            messages: {
                serializedName: "messages",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            }
        }
    }
};
export var ConfigServerSettingsValidateResult = {
    serializedName: "ConfigServerSettingsValidateResult",
    type: {
        name: "Composite",
        className: "ConfigServerSettingsValidateResult",
        modelProperties: {
            isValid: {
                serializedName: "isValid",
                type: {
                    name: "Boolean"
                }
            },
            details: {
                serializedName: "details",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ConfigServerSettingsErrorRecord"
                        }
                    }
                }
            }
        }
    }
};
export var ErrorModel = {
    serializedName: "Error",
    type: {
        name: "Composite",
        className: "ErrorModel",
        modelProperties: {
            code: {
                serializedName: "code",
                type: {
                    name: "String"
                }
            },
            message: {
                serializedName: "message",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var GitPatternRepository = {
    serializedName: "GitPatternRepository",
    type: {
        name: "Composite",
        className: "GitPatternRepository",
        modelProperties: {
            name: {
                required: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            pattern: {
                serializedName: "pattern",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            uri: {
                required: true,
                serializedName: "uri",
                type: {
                    name: "String"
                }
            },
            label: {
                serializedName: "label",
                type: {
                    name: "String"
                }
            },
            searchPaths: {
                serializedName: "searchPaths",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            username: {
                serializedName: "username",
                type: {
                    name: "String"
                }
            },
            password: {
                serializedName: "password",
                type: {
                    name: "String"
                }
            },
            hostKey: {
                serializedName: "hostKey",
                type: {
                    name: "String"
                }
            },
            hostKeyAlgorithm: {
                serializedName: "hostKeyAlgorithm",
                type: {
                    name: "String"
                }
            },
            privateKey: {
                serializedName: "privateKey",
                type: {
                    name: "String"
                }
            },
            strictHostKeyChecking: {
                serializedName: "strictHostKeyChecking",
                type: {
                    name: "Boolean"
                }
            }
        }
    }
};
export var ConfigServerGitProperty = {
    serializedName: "ConfigServerGitProperty",
    type: {
        name: "Composite",
        className: "ConfigServerGitProperty",
        modelProperties: {
            repositories: {
                serializedName: "repositories",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "GitPatternRepository"
                        }
                    }
                }
            },
            uri: {
                required: true,
                serializedName: "uri",
                type: {
                    name: "String"
                }
            },
            label: {
                serializedName: "label",
                type: {
                    name: "String"
                }
            },
            searchPaths: {
                serializedName: "searchPaths",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            username: {
                serializedName: "username",
                type: {
                    name: "String"
                }
            },
            password: {
                serializedName: "password",
                type: {
                    name: "String"
                }
            },
            hostKey: {
                serializedName: "hostKey",
                type: {
                    name: "String"
                }
            },
            hostKeyAlgorithm: {
                serializedName: "hostKeyAlgorithm",
                type: {
                    name: "String"
                }
            },
            privateKey: {
                serializedName: "privateKey",
                type: {
                    name: "String"
                }
            },
            strictHostKeyChecking: {
                serializedName: "strictHostKeyChecking",
                type: {
                    name: "Boolean"
                }
            }
        }
    }
};
export var ConfigServerSettings = {
    serializedName: "ConfigServerSettings",
    type: {
        name: "Composite",
        className: "ConfigServerSettings",
        modelProperties: {
            gitProperty: {
                serializedName: "gitProperty",
                type: {
                    name: "Composite",
                    className: "ConfigServerGitProperty"
                }
            }
        }
    }
};
export var ConfigServerProperties = {
    serializedName: "ConfigServerProperties",
    type: {
        name: "Composite",
        className: "ConfigServerProperties",
        modelProperties: {
            provisioningState: {
                readOnly: true,
                serializedName: "provisioningState",
                type: {
                    name: "String"
                }
            },
            error: {
                serializedName: "error",
                type: {
                    name: "Composite",
                    className: "ErrorModel"
                }
            },
            configServer: {
                serializedName: "configServer",
                type: {
                    name: "Composite",
                    className: "ConfigServerSettings"
                }
            }
        }
    }
};
export var ProxyResource = {
    serializedName: "ProxyResource",
    type: {
        name: "Composite",
        className: "ProxyResource",
        modelProperties: __assign({}, Resource.type.modelProperties)
    }
};
export var ConfigServerResource = {
    serializedName: "ConfigServerResource",
    type: {
        name: "Composite",
        className: "ConfigServerResource",
        modelProperties: __assign(__assign({}, ProxyResource.type.modelProperties), { properties: {
                serializedName: "properties",
                type: {
                    name: "Composite",
                    className: "ConfigServerProperties"
                }
            } })
    }
};
export var ApplicationInsightsAgentVersions = {
    serializedName: "applicationInsightsAgentVersions",
    type: {
        name: "Composite",
        className: "ApplicationInsightsAgentVersions",
        modelProperties: {
            java: {
                readOnly: true,
                serializedName: "java",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var MonitoringSettingProperties = {
    serializedName: "MonitoringSettingProperties",
    type: {
        name: "Composite",
        className: "MonitoringSettingProperties",
        modelProperties: {
            provisioningState: {
                readOnly: true,
                serializedName: "provisioningState",
                type: {
                    name: "String"
                }
            },
            error: {
                serializedName: "error",
                type: {
                    name: "Composite",
                    className: "ErrorModel"
                }
            },
            traceEnabled: {
                serializedName: "traceEnabled",
                type: {
                    name: "Boolean"
                }
            },
            appInsightsInstrumentationKey: {
                serializedName: "appInsightsInstrumentationKey",
                type: {
                    name: "String"
                }
            },
            appInsightsSamplingRate: {
                serializedName: "appInsightsSamplingRate",
                constraints: {
                    InclusiveMaximum: 100,
                    InclusiveMinimum: 0
                },
                type: {
                    name: "Number"
                }
            },
            appInsightsAgentVersions: {
                serializedName: "appInsightsAgentVersions",
                type: {
                    name: "Composite",
                    className: "ApplicationInsightsAgentVersions"
                }
            }
        }
    }
};
export var MonitoringSettingResource = {
    serializedName: "MonitoringSettingResource",
    type: {
        name: "Composite",
        className: "MonitoringSettingResource",
        modelProperties: __assign(__assign({}, ProxyResource.type.modelProperties), { properties: {
                serializedName: "properties",
                type: {
                    name: "Composite",
                    className: "MonitoringSettingProperties"
                }
            } })
    }
};
export var TestKeys = {
    serializedName: "TestKeys",
    type: {
        name: "Composite",
        className: "TestKeys",
        modelProperties: {
            primaryKey: {
                serializedName: "primaryKey",
                type: {
                    name: "String"
                }
            },
            secondaryKey: {
                serializedName: "secondaryKey",
                type: {
                    name: "String"
                }
            },
            primaryTestEndpoint: {
                serializedName: "primaryTestEndpoint",
                type: {
                    name: "String"
                }
            },
            secondaryTestEndpoint: {
                serializedName: "secondaryTestEndpoint",
                type: {
                    name: "String"
                }
            },
            enabled: {
                serializedName: "enabled",
                type: {
                    name: "Boolean"
                }
            }
        }
    }
};
export var RegenerateTestKeyRequestPayload = {
    serializedName: "RegenerateTestKeyRequestPayload",
    type: {
        name: "Composite",
        className: "RegenerateTestKeyRequestPayload",
        modelProperties: {
            keyType: {
                required: true,
                serializedName: "keyType",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var TemporaryDisk = {
    serializedName: "TemporaryDisk",
    type: {
        name: "Composite",
        className: "TemporaryDisk",
        modelProperties: {
            sizeInGB: {
                serializedName: "sizeInGB",
                constraints: {
                    InclusiveMaximum: 5,
                    InclusiveMinimum: 0
                },
                type: {
                    name: "Number"
                }
            },
            mountPath: {
                serializedName: "mountPath",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var PersistentDisk = {
    serializedName: "PersistentDisk",
    type: {
        name: "Composite",
        className: "PersistentDisk",
        modelProperties: {
            sizeInGB: {
                serializedName: "sizeInGB",
                constraints: {
                    InclusiveMaximum: 50,
                    InclusiveMinimum: 0
                },
                type: {
                    name: "Number"
                }
            },
            usedInGB: {
                readOnly: true,
                serializedName: "usedInGB",
                constraints: {
                    InclusiveMaximum: 50,
                    InclusiveMinimum: 0
                },
                type: {
                    name: "Number"
                }
            },
            mountPath: {
                serializedName: "mountPath",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var AppResourceProperties = {
    serializedName: "AppResourceProperties",
    type: {
        name: "Composite",
        className: "AppResourceProperties",
        modelProperties: {
            publicProperty: {
                serializedName: "public",
                type: {
                    name: "Boolean"
                }
            },
            url: {
                readOnly: true,
                serializedName: "url",
                type: {
                    name: "String"
                }
            },
            provisioningState: {
                readOnly: true,
                serializedName: "provisioningState",
                type: {
                    name: "String"
                }
            },
            activeDeploymentName: {
                serializedName: "activeDeploymentName",
                type: {
                    name: "String"
                }
            },
            fqdn: {
                serializedName: "fqdn",
                type: {
                    name: "String"
                }
            },
            httpsOnly: {
                serializedName: "httpsOnly",
                type: {
                    name: "Boolean"
                }
            },
            createdTime: {
                readOnly: true,
                serializedName: "createdTime",
                type: {
                    name: "DateTime"
                }
            },
            temporaryDisk: {
                serializedName: "temporaryDisk",
                type: {
                    name: "Composite",
                    className: "TemporaryDisk"
                }
            },
            persistentDisk: {
                serializedName: "persistentDisk",
                type: {
                    name: "Composite",
                    className: "PersistentDisk"
                }
            },
            enableEndToEndTLS: {
                serializedName: "enableEndToEndTLS",
                type: {
                    name: "Boolean"
                }
            }
        }
    }
};
export var AppResource = {
    serializedName: "AppResource",
    type: {
        name: "Composite",
        className: "AppResource",
        modelProperties: __assign(__assign({}, ProxyResource.type.modelProperties), { properties: {
                serializedName: "properties",
                type: {
                    name: "Composite",
                    className: "AppResourceProperties"
                }
            }, identity: {
                serializedName: "identity",
                type: {
                    name: "Composite",
                    className: "ManagedIdentityProperties"
                }
            }, location: {
                serializedName: "location",
                type: {
                    name: "String"
                }
            } })
    }
};
export var ResourceUploadDefinition = {
    serializedName: "ResourceUploadDefinition",
    type: {
        name: "Composite",
        className: "ResourceUploadDefinition",
        modelProperties: {
            relativePath: {
                serializedName: "relativePath",
                type: {
                    name: "String"
                }
            },
            uploadUrl: {
                serializedName: "uploadUrl",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var BindingResourceProperties = {
    serializedName: "BindingResourceProperties",
    type: {
        name: "Composite",
        className: "BindingResourceProperties",
        modelProperties: {
            resourceName: {
                readOnly: true,
                serializedName: "resourceName",
                type: {
                    name: "String"
                }
            },
            resourceType: {
                readOnly: true,
                serializedName: "resourceType",
                type: {
                    name: "String"
                }
            },
            resourceId: {
                serializedName: "resourceId",
                type: {
                    name: "String"
                }
            },
            key: {
                serializedName: "key",
                type: {
                    name: "String"
                }
            },
            bindingParameters: {
                serializedName: "bindingParameters",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "Object"
                        }
                    }
                }
            },
            generatedProperties: {
                readOnly: true,
                serializedName: "generatedProperties",
                type: {
                    name: "String"
                }
            },
            createdAt: {
                readOnly: true,
                serializedName: "createdAt",
                type: {
                    name: "String"
                }
            },
            updatedAt: {
                readOnly: true,
                serializedName: "updatedAt",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var BindingResource = {
    serializedName: "BindingResource",
    type: {
        name: "Composite",
        className: "BindingResource",
        modelProperties: __assign(__assign({}, ProxyResource.type.modelProperties), { properties: {
                serializedName: "properties",
                type: {
                    name: "Composite",
                    className: "BindingResourceProperties"
                }
            } })
    }
};
export var CertificateProperties = {
    serializedName: "CertificateProperties",
    type: {
        name: "Composite",
        className: "CertificateProperties",
        modelProperties: {
            thumbprint: {
                readOnly: true,
                serializedName: "thumbprint",
                type: {
                    name: "String"
                }
            },
            vaultUri: {
                required: true,
                serializedName: "vaultUri",
                type: {
                    name: "String"
                }
            },
            keyVaultCertName: {
                required: true,
                serializedName: "keyVaultCertName",
                type: {
                    name: "String"
                }
            },
            certVersion: {
                serializedName: "certVersion",
                type: {
                    name: "String"
                }
            },
            issuer: {
                readOnly: true,
                serializedName: "issuer",
                type: {
                    name: "String"
                }
            },
            issuedDate: {
                readOnly: true,
                serializedName: "issuedDate",
                type: {
                    name: "String"
                }
            },
            expirationDate: {
                readOnly: true,
                serializedName: "expirationDate",
                type: {
                    name: "String"
                }
            },
            activateDate: {
                readOnly: true,
                serializedName: "activateDate",
                type: {
                    name: "String"
                }
            },
            subjectName: {
                readOnly: true,
                serializedName: "subjectName",
                type: {
                    name: "String"
                }
            },
            dnsNames: {
                readOnly: true,
                serializedName: "dnsNames",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            }
        }
    }
};
export var CertificateResource = {
    serializedName: "CertificateResource",
    type: {
        name: "Composite",
        className: "CertificateResource",
        modelProperties: __assign(__assign({}, ProxyResource.type.modelProperties), { properties: {
                serializedName: "properties",
                type: {
                    name: "Composite",
                    className: "CertificateProperties"
                }
            } })
    }
};
export var NameAvailabilityParameters = {
    serializedName: "NameAvailabilityParameters",
    type: {
        name: "Composite",
        className: "NameAvailabilityParameters",
        modelProperties: {
            type: {
                required: true,
                serializedName: "type",
                type: {
                    name: "String"
                }
            },
            name: {
                required: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var NameAvailability = {
    serializedName: "NameAvailability",
    type: {
        name: "Composite",
        className: "NameAvailability",
        modelProperties: {
            nameAvailable: {
                serializedName: "nameAvailable",
                type: {
                    name: "Boolean"
                }
            },
            reason: {
                serializedName: "reason",
                type: {
                    name: "String"
                }
            },
            message: {
                serializedName: "message",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var CustomDomainProperties = {
    serializedName: "CustomDomainProperties",
    type: {
        name: "Composite",
        className: "CustomDomainProperties",
        modelProperties: {
            thumbprint: {
                serializedName: "thumbprint",
                type: {
                    name: "String"
                }
            },
            appName: {
                readOnly: true,
                serializedName: "appName",
                type: {
                    name: "String"
                }
            },
            certName: {
                serializedName: "certName",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var CustomDomainResource = {
    serializedName: "CustomDomainResource",
    type: {
        name: "Composite",
        className: "CustomDomainResource",
        modelProperties: __assign(__assign({}, ProxyResource.type.modelProperties), { properties: {
                serializedName: "properties",
                type: {
                    name: "Composite",
                    className: "CustomDomainProperties"
                }
            } })
    }
};
export var CustomDomainValidatePayload = {
    serializedName: "CustomDomainValidatePayload",
    type: {
        name: "Composite",
        className: "CustomDomainValidatePayload",
        modelProperties: {
            name: {
                required: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var CustomDomainValidateResult = {
    serializedName: "CustomDomainValidateResult",
    type: {
        name: "Composite",
        className: "CustomDomainValidateResult",
        modelProperties: {
            isValid: {
                serializedName: "isValid",
                type: {
                    name: "Boolean"
                }
            },
            message: {
                serializedName: "message",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ImageRegistryCredential = {
    serializedName: "ImageRegistryCredential",
    type: {
        name: "Composite",
        className: "ImageRegistryCredential",
        modelProperties: {
            username: {
                serializedName: "username",
                type: {
                    name: "String"
                }
            },
            password: {
                serializedName: "password",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var CustomContainer = {
    serializedName: "CustomContainer",
    type: {
        name: "Composite",
        className: "CustomContainer",
        modelProperties: {
            server: {
                serializedName: "server",
                type: {
                    name: "String"
                }
            },
            containerImage: {
                serializedName: "containerImage",
                type: {
                    name: "String"
                }
            },
            command: {
                serializedName: "command",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            args: {
                serializedName: "args",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            imageRegistryCredential: {
                serializedName: "imageRegistryCredential",
                type: {
                    name: "Composite",
                    className: "ImageRegistryCredential"
                }
            }
        }
    }
};
export var UserSourceInfo = {
    serializedName: "UserSourceInfo",
    type: {
        name: "Composite",
        className: "UserSourceInfo",
        modelProperties: {
            type: {
                serializedName: "type",
                type: {
                    name: "String"
                }
            },
            relativePath: {
                serializedName: "relativePath",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "version",
                type: {
                    name: "String"
                }
            },
            artifactSelector: {
                serializedName: "artifactSelector",
                type: {
                    name: "String"
                }
            },
            customContainer: {
                serializedName: "customContainer",
                type: {
                    name: "Composite",
                    className: "CustomContainer"
                }
            }
        }
    }
};
export var ResourceRequests = {
    serializedName: "ResourceRequests",
    type: {
        name: "Composite",
        className: "ResourceRequests",
        modelProperties: {
            cpu: {
                serializedName: "cpu",
                type: {
                    name: "String"
                }
            },
            memory: {
                serializedName: "memory",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DeploymentSettings = {
    serializedName: "DeploymentSettings",
    type: {
        name: "Composite",
        className: "DeploymentSettings",
        modelProperties: {
            cpu: {
                serializedName: "cpu",
                defaultValue: 1,
                type: {
                    name: "Number"
                }
            },
            memoryInGB: {
                serializedName: "memoryInGB",
                defaultValue: 1,
                type: {
                    name: "Number"
                }
            },
            resourceRequests: {
                serializedName: "resourceRequests",
                type: {
                    name: "Composite",
                    className: "ResourceRequests"
                }
            },
            jvmOptions: {
                serializedName: "jvmOptions",
                type: {
                    name: "String"
                }
            },
            netCoreMainEntryPath: {
                serializedName: "netCoreMainEntryPath",
                type: {
                    name: "String"
                }
            },
            environmentVariables: {
                serializedName: "environmentVariables",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            runtimeVersion: {
                serializedName: "runtimeVersion",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DeploymentInstance = {
    serializedName: "DeploymentInstance",
    type: {
        name: "Composite",
        className: "DeploymentInstance",
        modelProperties: {
            name: {
                readOnly: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            status: {
                readOnly: true,
                serializedName: "status",
                type: {
                    name: "String"
                }
            },
            reason: {
                readOnly: true,
                serializedName: "reason",
                type: {
                    name: "String"
                }
            },
            discoveryStatus: {
                readOnly: true,
                serializedName: "discoveryStatus",
                type: {
                    name: "String"
                }
            },
            startTime: {
                readOnly: true,
                serializedName: "startTime",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DeploymentResourceProperties = {
    serializedName: "DeploymentResourceProperties",
    type: {
        name: "Composite",
        className: "DeploymentResourceProperties",
        modelProperties: {
            source: {
                serializedName: "source",
                type: {
                    name: "Composite",
                    className: "UserSourceInfo"
                }
            },
            appName: {
                readOnly: true,
                serializedName: "appName",
                type: {
                    name: "String"
                }
            },
            deploymentSettings: {
                serializedName: "deploymentSettings",
                type: {
                    name: "Composite",
                    className: "DeploymentSettings"
                }
            },
            provisioningState: {
                readOnly: true,
                serializedName: "provisioningState",
                type: {
                    name: "String"
                }
            },
            status: {
                readOnly: true,
                serializedName: "status",
                type: {
                    name: "String"
                }
            },
            active: {
                readOnly: true,
                serializedName: "active",
                type: {
                    name: "Boolean"
                }
            },
            createdTime: {
                readOnly: true,
                serializedName: "createdTime",
                type: {
                    name: "DateTime"
                }
            },
            instances: {
                readOnly: true,
                serializedName: "instances",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "DeploymentInstance"
                        }
                    }
                }
            }
        }
    }
};
export var DeploymentResource = {
    serializedName: "DeploymentResource",
    type: {
        name: "Composite",
        className: "DeploymentResource",
        modelProperties: __assign(__assign({}, ProxyResource.type.modelProperties), { properties: {
                serializedName: "properties",
                type: {
                    name: "Composite",
                    className: "DeploymentResourceProperties"
                }
            }, sku: {
                serializedName: "sku",
                type: {
                    name: "Composite",
                    className: "Sku"
                }
            } })
    }
};
export var LogFileUrlResponse = {
    serializedName: "LogFileUrlResponse",
    type: {
        name: "Composite",
        className: "LogFileUrlResponse",
        modelProperties: {
            url: {
                required: true,
                serializedName: "url",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var OperationDisplay = {
    serializedName: "OperationDisplay",
    type: {
        name: "Composite",
        className: "OperationDisplay",
        modelProperties: {
            provider: {
                serializedName: "provider",
                type: {
                    name: "String"
                }
            },
            resource: {
                serializedName: "resource",
                type: {
                    name: "String"
                }
            },
            operation: {
                serializedName: "operation",
                type: {
                    name: "String"
                }
            },
            description: {
                serializedName: "description",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var LogSpecification = {
    serializedName: "LogSpecification",
    type: {
        name: "Composite",
        className: "LogSpecification",
        modelProperties: {
            name: {
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            displayName: {
                serializedName: "displayName",
                type: {
                    name: "String"
                }
            },
            blobDuration: {
                serializedName: "blobDuration",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var MetricDimension = {
    serializedName: "MetricDimension",
    type: {
        name: "Composite",
        className: "MetricDimension",
        modelProperties: {
            name: {
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            displayName: {
                serializedName: "displayName",
                type: {
                    name: "String"
                }
            },
            toBeExportedForShoebox: {
                serializedName: "toBeExportedForShoebox",
                type: {
                    name: "Boolean"
                }
            }
        }
    }
};
export var MetricSpecification = {
    serializedName: "MetricSpecification",
    type: {
        name: "Composite",
        className: "MetricSpecification",
        modelProperties: {
            name: {
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            displayName: {
                serializedName: "displayName",
                type: {
                    name: "String"
                }
            },
            displayDescription: {
                serializedName: "displayDescription",
                type: {
                    name: "String"
                }
            },
            unit: {
                serializedName: "unit",
                type: {
                    name: "String"
                }
            },
            category: {
                serializedName: "category",
                type: {
                    name: "String"
                }
            },
            aggregationType: {
                serializedName: "aggregationType",
                type: {
                    name: "String"
                }
            },
            supportedAggregationTypes: {
                serializedName: "supportedAggregationTypes",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            supportedTimeGrainTypes: {
                serializedName: "supportedTimeGrainTypes",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            fillGapWithZero: {
                serializedName: "fillGapWithZero",
                type: {
                    name: "Boolean"
                }
            },
            dimensions: {
                serializedName: "dimensions",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "MetricDimension"
                        }
                    }
                }
            }
        }
    }
};
export var ServiceSpecification = {
    serializedName: "ServiceSpecification",
    type: {
        name: "Composite",
        className: "ServiceSpecification",
        modelProperties: {
            logSpecifications: {
                serializedName: "logSpecifications",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "LogSpecification"
                        }
                    }
                }
            },
            metricSpecifications: {
                serializedName: "metricSpecifications",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "MetricSpecification"
                        }
                    }
                }
            }
        }
    }
};
export var OperationProperties = {
    serializedName: "OperationProperties",
    type: {
        name: "Composite",
        className: "OperationProperties",
        modelProperties: {
            serviceSpecification: {
                serializedName: "serviceSpecification",
                type: {
                    name: "Composite",
                    className: "ServiceSpecification"
                }
            }
        }
    }
};
export var OperationDetail = {
    serializedName: "OperationDetail",
    type: {
        name: "Composite",
        className: "OperationDetail",
        modelProperties: {
            name: {
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            isDataAction: {
                serializedName: "isDataAction",
                type: {
                    name: "Boolean"
                }
            },
            display: {
                serializedName: "display",
                type: {
                    name: "Composite",
                    className: "OperationDisplay"
                }
            },
            origin: {
                serializedName: "origin",
                type: {
                    name: "String"
                }
            },
            properties: {
                serializedName: "properties",
                type: {
                    name: "Composite",
                    className: "OperationProperties"
                }
            }
        }
    }
};
export var SkuCapacity = {
    serializedName: "SkuCapacity",
    type: {
        name: "Composite",
        className: "SkuCapacity",
        modelProperties: {
            minimum: {
                required: true,
                serializedName: "minimum",
                type: {
                    name: "Number"
                }
            },
            maximum: {
                serializedName: "maximum",
                type: {
                    name: "Number"
                }
            },
            default: {
                serializedName: "default",
                type: {
                    name: "Number"
                }
            },
            scaleType: {
                serializedName: "scaleType",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ResourceSkuCapabilities = {
    serializedName: "ResourceSkuCapabilities",
    type: {
        name: "Composite",
        className: "ResourceSkuCapabilities",
        modelProperties: {
            name: {
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            value: {
                serializedName: "value",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ResourceSkuZoneDetails = {
    serializedName: "ResourceSkuZoneDetails",
    type: {
        name: "Composite",
        className: "ResourceSkuZoneDetails",
        modelProperties: {
            name: {
                serializedName: "name",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            capabilities: {
                serializedName: "capabilities",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ResourceSkuCapabilities"
                        }
                    }
                }
            }
        }
    }
};
export var ResourceSkuLocationInfo = {
    serializedName: "ResourceSkuLocationInfo",
    type: {
        name: "Composite",
        className: "ResourceSkuLocationInfo",
        modelProperties: {
            location: {
                serializedName: "location",
                type: {
                    name: "String"
                }
            },
            zones: {
                serializedName: "zones",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            zoneDetails: {
                serializedName: "zoneDetails",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ResourceSkuZoneDetails"
                        }
                    }
                }
            }
        }
    }
};
export var ResourceSkuRestrictionInfo = {
    serializedName: "ResourceSkuRestrictionInfo",
    type: {
        name: "Composite",
        className: "ResourceSkuRestrictionInfo",
        modelProperties: {
            locations: {
                serializedName: "locations",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            zones: {
                serializedName: "zones",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            }
        }
    }
};
export var ResourceSkuRestrictions = {
    serializedName: "ResourceSkuRestrictions",
    type: {
        name: "Composite",
        className: "ResourceSkuRestrictions",
        modelProperties: {
            type: {
                serializedName: "type",
                type: {
                    name: "String"
                }
            },
            values: {
                serializedName: "values",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            restrictionInfo: {
                serializedName: "restrictionInfo",
                type: {
                    name: "Composite",
                    className: "ResourceSkuRestrictionInfo"
                }
            },
            reasonCode: {
                serializedName: "reasonCode",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ResourceSku = {
    serializedName: "ResourceSku",
    type: {
        name: "Composite",
        className: "ResourceSku",
        modelProperties: {
            resourceType: {
                serializedName: "resourceType",
                type: {
                    name: "String"
                }
            },
            name: {
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            tier: {
                serializedName: "tier",
                type: {
                    name: "String"
                }
            },
            capacity: {
                serializedName: "capacity",
                type: {
                    name: "Composite",
                    className: "SkuCapacity"
                }
            },
            locations: {
                serializedName: "locations",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            locationInfo: {
                serializedName: "locationInfo",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ResourceSkuLocationInfo"
                        }
                    }
                }
            },
            restrictions: {
                serializedName: "restrictions",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ResourceSkuRestrictions"
                        }
                    }
                }
            }
        }
    }
};
export var SupportedRuntimeVersion = {
    serializedName: "SupportedRuntimeVersion",
    type: {
        name: "Composite",
        className: "SupportedRuntimeVersion",
        modelProperties: {
            value: {
                serializedName: "value",
                type: {
                    name: "String"
                }
            },
            platform: {
                serializedName: "platform",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "version",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var AvailableRuntimeVersions = {
    serializedName: "AvailableRuntimeVersions",
    type: {
        name: "Composite",
        className: "AvailableRuntimeVersions",
        modelProperties: {
            value: {
                readOnly: true,
                serializedName: "value",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "SupportedRuntimeVersion"
                        }
                    }
                }
            }
        }
    }
};
export var ServiceResourceList = {
    serializedName: "ServiceResourceList",
    type: {
        name: "Composite",
        className: "ServiceResourceList",
        modelProperties: {
            value: {
                serializedName: "",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ServiceResource"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var AppResourceCollection = {
    serializedName: "AppResourceCollection",
    type: {
        name: "Composite",
        className: "AppResourceCollection",
        modelProperties: {
            value: {
                serializedName: "",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "AppResource"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var BindingResourceCollection = {
    serializedName: "BindingResourceCollection",
    type: {
        name: "Composite",
        className: "BindingResourceCollection",
        modelProperties: {
            value: {
                serializedName: "",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "BindingResource"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var CertificateResourceCollection = {
    serializedName: "CertificateResourceCollection",
    type: {
        name: "Composite",
        className: "CertificateResourceCollection",
        modelProperties: {
            value: {
                serializedName: "",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "CertificateResource"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var CustomDomainResourceCollection = {
    serializedName: "CustomDomainResourceCollection",
    type: {
        name: "Composite",
        className: "CustomDomainResourceCollection",
        modelProperties: {
            value: {
                serializedName: "",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "CustomDomainResource"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DeploymentResourceCollection = {
    serializedName: "DeploymentResourceCollection",
    type: {
        name: "Composite",
        className: "DeploymentResourceCollection",
        modelProperties: {
            value: {
                serializedName: "",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "DeploymentResource"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var AvailableOperations = {
    serializedName: "AvailableOperations",
    type: {
        name: "Composite",
        className: "AvailableOperations",
        modelProperties: {
            value: {
                serializedName: "",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "OperationDetail"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ResourceSkuCollection = {
    serializedName: "ResourceSkuCollection",
    type: {
        name: "Composite",
        className: "ResourceSkuCollection",
        modelProperties: {
            value: {
                serializedName: "",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ResourceSku"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
