// @azure/arm-appplatform/esm/models/parameters.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
export var acceptLanguage = {
    parameterPath: "acceptLanguage",
    mapper: {
        serializedName: "accept-language",
        defaultValue: 'en-US',
        type: {
            name: "String"
        }
    }
};
export var apiVersion = {
    parameterPath: "apiVersion",
    mapper: {
        required: true,
        serializedName: "api-version",
        constraints: {
            MinLength: 1
        },
        type: {
            name: "String"
        }
    }
};
export var appName = {
    parameterPath: "appName",
    mapper: {
        required: true,
        serializedName: "appName",
        type: {
            name: "String"
        }
    }
};
export var bindingName = {
    parameterPath: "bindingName",
    mapper: {
        required: true,
        serializedName: "bindingName",
        type: {
            name: "String"
        }
    }
};
export var certificateName = {
    parameterPath: "certificateName",
    mapper: {
        required: true,
        serializedName: "certificateName",
        type: {
            name: "String"
        }
    }
};
export var deploymentName = {
    parameterPath: "deploymentName",
    mapper: {
        required: true,
        serializedName: "deploymentName",
        type: {
            name: "String"
        }
    }
};
export var domainName = {
    parameterPath: "domainName",
    mapper: {
        required: true,
        serializedName: "domainName",
        type: {
            name: "String"
        }
    }
};
export var location = {
    parameterPath: "location",
    mapper: {
        required: true,
        serializedName: "location",
        type: {
            name: "String"
        }
    }
};
export var nextPageLink = {
    parameterPath: "nextPageLink",
    mapper: {
        required: true,
        serializedName: "nextLink",
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export var resourceGroupName = {
    parameterPath: "resourceGroupName",
    mapper: {
        required: true,
        serializedName: "resourceGroupName",
        type: {
            name: "String"
        }
    }
};
export var serviceName = {
    parameterPath: "serviceName",
    mapper: {
        required: true,
        serializedName: "serviceName",
        type: {
            name: "String"
        }
    }
};
export var subscriptionId = {
    parameterPath: "subscriptionId",
    mapper: {
        required: true,
        serializedName: "subscriptionId",
        type: {
            name: "String"
        }
    }
};
export var syncStatus = {
    parameterPath: [
        "options",
        "syncStatus"
    ],
    mapper: {
        serializedName: "syncStatus",
        type: {
            name: "String"
        }
    }
};
export var version = {
    parameterPath: [
        "options",
        "version"
    ],
    mapper: {
        serializedName: "version",
        type: {
            name: "Sequence",
            element: {
                type: {
                    name: "String"
                }
            }
        }
    },
    collectionFormat: msRest.QueryCollectionFormat.Multi
};
